### PR TITLE
[prometheus] Change resources determination for VPA

### PR DESCRIPTION
**Pull request by Diaderius**
Signed-off-by: Konstantin Maltsev <konstantin.maltsev@flant.com>

## Description
There is a fix for detect_vpa_max hook for Prometheus resources configuration.

## Why do we need it, and what problem does it solve?
There is a configuration of minimal resources (minMem and minCPU) for Prometheus, which equals to **1000Mi** and **200m** accordingly. For Prometheus longterm **500Mi** and **50m** accordingly.
It refers to this issue: #3308 

## What is the expected result?
If minMem is more than maxMem, then make them equal. The same for minCPU and maxCPU.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: prometheus
type: fix 
summary: change resources determination for Prometheus
```

### CODE REVIEW

File: modules/300-prometheus/openapi/config-values.yaml [hunk 1]
```diff
@@ -167,6 +167,26 @@ properties:
...
+      minMemory:
+        oneOf:
+          - type: string
+            pattern: '^[0-9]+(\.[0-9]+)?(E|P|T|G|M|k|Ei|Pi|Ti|Gi|Mi|Ki)?$'
+          - type: number
+        x-examples: [ "3Mi" ]
+        description: |
+          The minimum Memory requests set by the VPA controller for the main Prometheus Pods.
+
+          The default value is chosen automatically based on the minimum number of resources for prometheus work.
```
Comment (yalosev) [inline modules/300-prometheus/openapi/config-values.yaml:189]: Nope. We don't need these parameters to be configured by user.

File: modules/300-prometheus/templates/prometheus/prometheus.yaml [hunk 2]
```diff
@@ -27,7 +22,8 @@ spec:
     containerPolicies:
     - containerName: "prometheus"
       minAllowed:
-        {{- include "prometheus_resources" . | nindent 8 }}
+        cpu: { { .Values.prometheus.vpa.minCPU | default .Values.prometheus.internal.vpa.minCPU | quote } }
+        memory: { { .Values.prometheus.vpa.minMemory | default .Values.prometheus.internal.vpa.minMemory | quote } }
```
Comment (yalosev) [inline modules/300-prometheus/templates/prometheus/prometheus.yaml:26]: also rollback this

File: modules/300-prometheus/templates/prometheus/prometheus.yaml [hunk 3]
```diff
@@ -1,8 +1,3 @@
-{{- define "prometheus_resources" }}
-cpu: 200m
-memory: 1000Mi
-{{- end }}
```
Comment (yalosev) [inline modules/300-prometheus/templates/prometheus/prometheus.yaml:4]: don't delete it. Just hardcode it into the hook

File: modules/300-prometheus/hooks/detect_vpa_max.go [hunk 4]
```diff
@@ -84,13 +84,26 @@ func calculateNodesCapacity(input *go_hook.HookInput) error {
 	totalPodsMemory := totalPodsCapacity * memoryPerPod
 	totalPodsCPU := totalPodsCapacity * milliCPUPerPod
 
-	// calculate prometheus usage
-	maxMem := resource.NewQuantity(totalPodsMemory/1, resource.BinarySI)
-	maxCPU := resource.NewMilliQuantity(totalPodsCPU/1, resource.DecimalSI)
+	// calculate prometheus usage by MustParse method
+	minMem := resource.MustParse("1000Mi")
+	maxMem := resource.MustParse(fmt.Sprintf("%dMi", totalPodsMemory/1))
```
Comment (yalosev) [inline modules/300-prometheus/hooks/detect_vpa_max.go:89]: Why have you changed it? MustParse is slower and more dangerous. We can left maxMem and maxCPU as they were

File: modules/300-prometheus/hooks/detect_vpa_max.go [hunk 5]
```diff
@@ -84,13 +84,26 @@ func calculateNodesCapacity(input *go_hook.HookInput) error {
 	totalPodsMemory := totalPodsCapacity * memoryPerPod
 	totalPodsCPU := totalPodsCapacity * milliCPUPerPod
 
-	// calculate prometheus usage
-	maxMem := resource.NewQuantity(totalPodsMemory/1, resource.BinarySI)
-	maxCPU := resource.NewMilliQuantity(totalPodsCPU/1, resource.DecimalSI)
+	// calculate prometheus usage by MustParse method
+	minMem := resource.MustParse("1000Mi")
```
Comment (yalosev) [inline modules/300-prometheus/hooks/detect_vpa_max.go:88]: We don't want to calculate it the every iteration. Just put it to the `var` block as `resource.NewQuantity...`

**Review by yalosev — CHANGES_REQUESTED** [review]

File: modules/300-prometheus/hooks/detect_vpa_max.go [hunk 6]
```diff
@@ -88,10 +90,21 @@ func calculateNodesCapacity(input *go_hook.HookInput) error {
...
+
+	if maxCPU.Cmp(*minCPU) == -1 {
+		maxCPU = minCPU
+	}
+
+	// calculate longterm prometheus usage
 	maxLongtermMem := resource.NewQuantity(totalPodsMemory/3, resource.BinarySI)
 	maxLongtermCPU := resource.NewMilliQuantity(totalPodsCPU/3, resource.DecimalSI)
 
+	input.Values.Set("prometheus.internal.vpa.minMemory", minMem.String())
```
Comment (yalosev) [inline modules/300-prometheus/hooks/detect_vpa_max.go:105]: What for do we need this value?

Comment (Diaderius) [reply modules/300-prometheus/hooks/detect_vpa_max.go]: @yalosev my bad, missed it from previous logic.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Diaderius. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Diaderius. I've started the workflow for label 'e2e/run/aws'...

**Review by yalosev — APPROVED** [review]

**Pull request merged**